scapy/contrib/coap_socket.py

```diff
@@ -0,0 +1,1001 @@
+    >>>         # You can handle the packages inside your resources,
+    >>>         # here will only be the "unhandled" ones.
+
+    :param ip: ip address to bind udp socket to.
+    :param port: port to bind udp socket to.
+    :param ack_timeout: the time, in ms, that we should wait for the acknowledgment
+                        after sending a request.
+    :param retries: amount of retransmissions before giving up on the request.
+    :param duplication_response_timeout: Timeout, in fractions of seconds, that we will
+                                        keep the response in case a response get lost.
+    :param lst_resources: optional, list of registered resources.
+    :param sock: optional, a socket instance to transmit,
+                 if None, a classic UDP socket will be open and bound to ip/port.
+    :param close_on_timeout: Will try to close the socket if the retries is exceeded
+    """
+
+    def __init__(self,
+                 ip="",  # type: str
+                 port=5683,  # type: int
+                 ack_timeout=500,  # type: int
+                 retries=3,  # type: int
+                 duplication_response_timeout=1.00,  # type: float
+                 lst_resources=None,  # type: Optional[None, list[CoAPResource]]
+                 sock=None,  # type: Optional[None, SuperSocket, any]
+                 close_on_timeout=False  # type: bool
+                 ):
+        self.impl = CoAPSocketImpl(ip, port, ack_timeout, retries,
+                                   duplication_response_timeout, lst_resources, sock,
+                                   close_on_timeout)
+
+        self.ins = cast(socket.socket, self.impl)
+        self.outs = cast(socket.socket, self.impl)
+        self.basecls = CoAP
+
+    def recv_raw(self, x=0xffff):
+        # type: (int) -> Tuple[Optional[Type[Packet]], Optional[bytes], Optional[float]]
+        if not self.closed:
+            tup = self.impl.recv()
+            if tup is not None:
+                return IP, tup[0], float(tup[1])
+        return IP, None, None
+
+    def recv(self, x=MTU, **kwargs):
+        # type: (int, **Any) -> Optional[Packet]
+        return super(CoAPSocket, self).recv(x, **kwargs)
+
+    def close(self):
+        # type: () -> None
+        if not self.closed:
+            self.closed = True
+            self.impl.close()
+
+    def send(self, x):
+        # type: (Packet) -> int
+        """
+        Send the packet using this socket.
+        Should be a CoAP packet with IP and UDP data.
+
+        Example:
+        >>> IP(dst="192.168.1.1") / UDP(dport=1234) / CoAP()
+        >>> IP(dst="192.168.1.1") / UDP(dport=1234) / CoAPSocket.make_coap_req_packet()
+
+        :param x: Concatenated packet with IP / UDP / CoAP
+        :return: The length of x, which is the amount of bytes sent
+        """
+        self.impl.send(x.dst, x.dport, x[CoAP])
+        return len(x)
+
+    def sr(self, *args, **kargs):
+        args[0].sport = self.impl.port
```

Comment:
since you want to operate on the packet you get here, I suggest to change the function signature:
```
def sr(pkt, *args, **kargs):
    pkt[UDP].sport = ...
    ...

Reply:
Hopefully I got this correct, did you mean:

```
def sr(self, pkt, *args, **kargs):
    pkt[UDP] ...
```

Reply:
Looks good. Unfortunately I’m not fully done with the review. I try to find some time as soon as possible.